### ISSUE TITLE
Redundant closure lint recommends incorrect code

### ISSUE THREAD

**Issue by iajoiner** [rust-lang/rust-clippy#10378]
### Summary

This is probably a variant of https://github.com/rust-lang/rust-clippy/issues/3071 or https://github.com/rust-lang/rust-clippy/issues/7435. This error is probably dependent on PyO3 which is why I have filed an issue there as well as https://github.com/PyO3/pyo3/issues/2988.

The reason why it is necessary to file an issue here is that the suggestion doesn't actually fix the issue. In fact the same issue exists after refactoring the code according to the suggestion. The next suggestion is for us to directly use Vec::new as a function which causes an error.

### Lint Name

redundant-closure

### Reproducer

I tried this code:

```rust
use pyo3::prelude::*;

#[pyfunction]
#[pyo3(signature = (
    _data = vec![],
))]
pub fn foo(
    _data: Vec<i32>,
) {
}
```

I saw this happen:

```
Checking pycatch v0.1.0 (/home/iajoiner/Documents/tests/pycatch)
warning: redundant closure
 --> src/main.rs:5:13
  |
5 |       _data = vec![],
  |  _____________^
6 | | ))]
7 | | pub fn foo(
8 | |     _data: Vec<i32>,
  | |___________________^ help: replace the closure with `Vec::new`: `std::vec::Vec::new`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
  = note: `#[warn(clippy::redundant_closure)]` on by default

warning: `pycatch` (bin "pycatch") generated 1 warning (run `cargo clippy --fix --bin "pycatch"` to apply 1 suggestion)
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
```

I expected to see this happen:

### Version

```text
rustc 1.67.1 (d5a82bbd2 2023-02-07)
binary: rustc
commit-hash: d5a82bbd26e1ad8b7401f6a718a9c57c96905483
commit-date: 2023-02-07
host: x86_64-unknown-linux-gnu
release: 1.67.1
LLVM version: 15.0.6
```

### Additional Labels

-  `I-suggestion-causes-error`

**Comment by iajoiner** [comment]
Minimal example added. This is a weird error that is preventing Python bindings of Apache Arrow Datafusion from using clippy in our CI.

**Comment by iajoiner** [comment]
https://github.com/PyO3/pyo3/issues/2988

**Comment by iajoiner** [comment]
Closed as this is not a Clippy issue. It is a PyO3 one.

**Comment by davidhewitt** [comment]
Despite having fixed this in PyO3, I would be interested to hear if there are techniques to hint clippy that the PyO3 code is generated and users can't easily correct it.
